sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java

```diff
@@ -636,4 +648,49 @@ public void visitPrimitiveTransform(Node node) {
       node.getTransform().validate(options);
     }
   }
+
+  private static class TransformToMessage implements Function<PTransform<?, ?>, String> {
+    private static final TransformToMessage INSTANCE = new TransformToMessage();
+
+    @Override
+    public String apply(final PTransform<?, ?> transform) {
+      return "    - " + transform;
+    }
+  }
+
+  private static class UnstableNameToMessage implements
+          Function<Map.Entry<String, Collection<PTransform<?, ?>>>, String> {
+    private final Multimap<String, PTransform<?, ?>> instances;
+
+    private UnstableNameToMessage(final Multimap<String, PTransform<?, ?>> instancePerName) {
+      this.instances = instancePerName;
+    }
+
+    @Override
+    public String apply(final Map.Entry<String, Collection<PTransform<?, ?>>> input) {
+      final Collection<PTransform<?, ?>> values = instances.get(input.getKey());
+      return "- name=" + input.getKey() + ":\n"
+              + Joiner.on("\n").join(transform(values, TransformToMessage.INSTANCE));
+    }
+  }
+
+  private static class KeysExtractor implements
+          Function<Map.Entry<String, Collection<PTransform<?, ?>>>, String> {
+    private static final KeysExtractor INSTANCE = new KeysExtractor();
```

Comment:
These INSTANCE variables are probably not worth the extra lines of code: just use `new KeysExtractor()` and `new IsUnique()`